### DISCUSSION TITLE
Empty Send MSQUIC Server

### DISCUSSION THREAD

**Discussion by JyothiAswath** [microsoft/msquic#4550] [Q&A]
Hi,
MSQUIC Version - v2.2.4

A sample application is continuously sending data from the msquic server to the msquic client. At different times in the sequence it is noticed that though the right data has been sent from the MSQUIC Server ( MsQuic->StreamSend API call), the client  receives an empty buffer. 
After enabling the debug logs on the MSQUIC Server and adding some more debug statements it is observed that the data which is sent is present while the control is still in api.c on the server. In the server itself when the control reaches stream_send.c the data is all set to empty (0x0 in Hexa for the exact length of the data which was sent)

**Log Excerpt on MSQUIC Server:**
**At this point 44 bytes has exact data which was intended to be sent.**
[strm][0x7fc10c03edf0] App queuing send [44 bytes, 1 buffers, 0x1 flags] [StreamAppSend:/msquic/src/core/api.c:1069]

**At this point 44 bytes had just an empty buffer (0x0 in Hexa)**
Send Request [0x7fc104001df0] queued with 44 bytes at offset 4446 (flags 0x1) [SendQueued:/msquic/src/core/stream_send.c:566]

Any guidance and help would be greatly appreciated

**Comment by JyothiAswath** [comment]
One more key observation -
The buffer has valid contents when enqueued - QuicOperationEnqueue( connection.c) . The buffer has empty contents after it is dequeued - QuicConnDrainOperations.

The same is resulting with MSQUIC version - v2.3.6
@nibanks any help here for the way forward.

**Reply by nibanks** [reply]
That seems to indicate you are changing the buffer before you get a send complete. Just because `StreamSend` returns doesn't mean it's 'done'. It returns `QUIC_STATUS_PENDING`.

**Reply by JyothiAswath** [reply]
This is working. Appreciate your help.. In case I need any support after testing some more time, will get back here.
